pkg/indexbuilder/ottawau.py

```diff
@@ -0,0 +1,81 @@
+import yaml
+import re
+import os
+from os import path
+from collections import defaultdict
+
+import nltk
+from .indexbuilder import IndexBuilder
+from .invertedindex import IndexKey
+from ..dictionary import Dictionary, DictBuilder
+from ..wordmodifiers import *
+
+
+class OttawaUIndexBuilder(IndexBuilder):
+    def __init__(self, ctx):
+        self.corpus_path = ctx.corpus_path
+        self.dict_path = ctx.dict_path
+        self.inverted_index_path = ctx.inverted_index_path
+        self.tokenizer = ctx.tokenizer
+
+        # TODO extract filter func construction to shareable util module that takes as argument a Context object
+        # normalizing functions must be called directly on the corpus, not the tokenized terms
+        # note that normalizing function change the actual text of the corpus, not a generated token (which happens further downstream)
+        self.normalize_funcs = []
+        if ctx.enable_normalization:
+            self.normalize_funcs.append(Normalizer.normalize_periods)
+            self.normalize_funcs.append(Normalizer.normalize_hyphens)
+
+        # filter functions should be strictly functional/idempotent and take as parameters only (self, set)
+        self.filter_funcs = []
+        if ctx.enable_casefolding:
+            self.filter_funcs.append(CaseFolder().call)
+        if ctx.enable_stopwords:
+            self.filter_funcs.append(StopWordFilter(DictBuilder.BASE_STOPWORDS).call)
+        if ctx.remove_nonalphanumeric:
+            self.filter_funcs.append(AlphaNumericFilter().call)
+        if ctx.enable_stemming:
+            self.filter_funcs.append(Stemmer(nltk.LancasterStemmer()).call)
+
+    def build(self):
+        term_documents_dict = self._build_simple_index()
+
+        inverted_index = {}
+        for key in term_documents_dict:
+            doc_ids = term_documents_dict[key]
+            key_with_freq = IndexKey(key, len(doc_ids))
+            inverted_index[key_with_freq] = doc_ids
+
```

Comment:
The `simple_index` is just a mapping of terms to a list of document IDs. E.g. if the term "foobar" occurs in documents 2,3, and 4, the resulting entry will be:
`"foobar": [2,3,4]"`

Following this, we create the actual inverted index by adding the document frequencies to the keys of the set. Using the above example, we end up with:
`foobar (3): [2,3,4]`